Added /condense command that converts all items into blocks.

FarisR99 opened the pull request:
Adds a new command (/block) that converts all items into block forms (if possible). Made sure to add the messages but only in English and default. I believe this will be useful as an "essential" command as it makes life easier for players and can be used as donor perks which Essential commands are used for a lot (e.g. /hat)

FarisR99 committed:
Added /block command that converts all items into blocks.

khobbits commented:
I was thinking of adding something similar to this called /condense, which would do effectively do the same thing, but in theory not be limited to just item->block, but there are no good examples in vanilla for this atm.

If we are adding this, we'd probably need to make it automatic, to support new block types, and mods, so would need to use recipes rather than a hard coded mapping.

khobbits commented on Essentials/src/com/earth2me/essentials/commands/Commandblock.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.earth2me.essentials.I18n._;
```
This is the old tl system, doesnt exist in 2.x now, you might want to update your 2.x branch.

FarisR99 committed:
Updated to the new 2.x

Updated to the new 2.x Essentials branch. My Git wasn't working properly for some reason and refused to update.

FarisR99 commented:
So to allow modification, do you allow me to make a new EssentialsConf for blocks.yml?

Edit: Never mind, I could just check if the recipe ingredients fills up all 9x9 and all consists of the same material type.

FarisR99 committed:
Detects custom recipes and not hard coded.

Detects custom recipes and not hard coded into a Map for the list of materials.

Note: This has not been tested but it seems legit.

OscarWoHA commented:
Why not just use the items.csv-file?

FarisR99 commented:
The items.csv has no relation to the block list as items.csv is just an alias base is it not?

FarisR99 committed:
Made code more efficient and bug-proof

Made the code more efficient in terms of performance by only getting variables when necessary. Also, made sure the inventory does not bug for the player as removing items in a loop can cause inventories to glitch sometimes.

FarisR99 committed:
Updated message for multi-purposes

Allowed the "itemsNotConverted" to be used not only for not being converted but also for not having enough in the inventory.

chrisgward commented:
Your code does not compile - missing semicolon

Essentials/Essentials/src/com/earth2me/essentials/commands/Commandblock.java:[112,36] error: ';' expected

FarisR99 committed:
Fixed missing semi-colon.

necrodoom commented on Essentials/src/com/earth2me/essentials/commands/Commandblock.java:
```diff
@@ -0,0 +1,120 @@
+	@Override
+	protected void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+		if (args.length == 0) {
+			if (user.isAuthorized("essentials.block")) {
```
remove this check, essentials already does it

necrodoom commented on Essentials/src/messages.properties:
```diff
@@ -190,7 +190,10 @@ itemSold=\u00a7aSold for \u00a7c{0} \u00a7a({1} {2} at {3} each).
 itemSoldConsole=\u00a7a{0} \u00a7asold {1} for \u00a7a{2} \u00a7a({3} items at {4} each).
 itemSpawn=\u00a76Giving\u00a7c {0} \u00a76of\u00a7c {1}
 itemType=\u00a76Item\:\u00a7c {0} \u00a76-\u00a7c {1}
+itemsConverted=\u00a76Converted all items into blocks.
```
copy-paste these to all other locales

FarisR99 replied:
You would like me to translate every single one and add them to every single locale? :/
Or should I just add these in English to every other locale.

khobbits replied:
The lines just need copy pasting to each locale, unless you are a native speaker of the language, leave them in English.

FarisR99 committed:
Removed permission check and changed user.getBase() into player.

FarisR99 committed:
Added item messages.

khobbits commented:
The logic on this command seems overly complicated for what it needs to do, i'll look into this and see if it is necessary.

FarisR99 commented:
Okay. =)

FarisR99 committed:
Reverted itemsCsvNotLoaded back.

Reverted itemsCsvNotLoaded back to the correct translation.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Updated messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Update messages_ro.properties

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 committed:
Added item messages.

FarisR99 commented on Essentials/src/com/earth2me/essentials/commands/Commandblock.java:
```diff
@@ -0,0 +1,117 @@
+		List<Recipe> recipeList = new ArrayList<Recipe>();
+		while (recipeIter.hasNext())
+			recipeList.add(recipeIter.next());
+		for (Recipe recipe : recipeList) {
```
Does not work if called when the plugin is loading. Must be called once the server is loaded.

FarisR99 replied:
Is there a new instance of this class created every time the command is run? If called in onEnable(), should I make a scheduled task for a 1 second delay?

FarisR99 commented on Essentials/src/com/earth2me/essentials/commands/Commandblock.java:
```diff
@@ -0,0 +1,117 @@
+		for (Recipe recipe : recipeList) {
+			if (recipe instanceof ShapedRecipe) {
+				ShapedRecipe shapedRecipe = (ShapedRecipe) recipe;
+				if (shapedRecipe.getResult() != null && shapedRecipe.getResult().getType().isBlock()) {
```
Could just change this into: if (shapedRecipe.getResult().toString().endsWith("_BLOCK"))?

FarisR99 committed:
Fixed the failure of checking the item stacks.

Fixed the failure of checking the recipe results by setting isFirst to true by default.

FarisR99 committed:
Fully updated code

Full updated code that has been tested.

FarisR99 committed:
Updated to my coding style.

khobbits commented:
Essentials commands are instanced every time they are used, i'm not sure this model works.
We will not be preloading commands, it goes against our command model.

FarisR99 commented:
Then this does work, it is just inefficient in terms of adding to the map every time. I'll make it static and then run a delayed task to intialise them or make it so in the constructor, if the Map is null (by default, null), initialise.

FarisR99 committed:
Changed to the more preferred way of sending messages.

khobbits commented:
Just thought I'd play with the idea for a bit, this is what I came up with.
https://gist.github.com/khobbits/3a4a6956b02e51be3dec

khobbits commented:
Whoops, another tiny update, this time added support for /condense [hand|all|[itemname]]

This version doesn't enumerate recipes which aren't used, and doesn't break the command model of doing early initialization.

This solution also doesn't convert any unconvertable stacks by default, so it won't convert snowballs to snowblocks, unless you use /condense snowball or /condense all.

FarisR99 commented:
Wow, that seems good but it seems a bit too complicated and some bits may not required I think something like this to add support must be that complicated though. I'll test your code and do some performance checks before committing if that is alright with you. With my old code there was a slight problem that may occur in yours too. If every diamond stack had some remainders, and the diamond stack left over was greater than 8, it wouldn't condense the left over.

FarisR99 committed:
Temporarily comitted KHobbit's code

Updated to KHobbit's code.
Note: Not tested - will do when out of school. Also will need formatting.

khobbits commented:
Mine shouldn't have that problem, since I total up all the items of the same type first and combine them.

Mine was getting a little complicated, so I split it up into methods, which bloats the code size a little bit.
I also used the Essentials code convention, which requires brackets on the next line.  That means there is quite a bit of 'line bloat'.

The 'SimpleRecipe' class at the end was to just try and simplify the map stores.

The 'validateReverse' stuff I think is quite important.  If someone types /condense snowball, or /condense all, I don't see any reason not to convert snowballs to snowblocks, but this 'condensing' is a one way process, you can't simply condense.  The same goes for goldnuggets to goldingots.

For this reason, I think the default state should be to only condense blocks which can be unconverted.

FarisR99 committed:
Tested and working code

Testing and working code pushed.

FarisR99 commented:
Ah okay, yeah I pushed the code, sorry, my preference is for the {}'s to be on the same line and so it's back to that.
This has been tested by me on a LAN server and works perfectly fine.
Following things tested:
- Typing the command while having one stack of convert-able items.
- Typing the command while having multiple different convert-able items.
- Typing the command while having the first slot with an inconvertible item.
- Typing the command while having the first slot with an inconvertible item and multiple different convert-able items.

FarisR99 committed:
Removed surpress warning and attempted to succeed build

khobbits commented:
Code that doesn't match codestyle requirements will never be directly merged:
- Codestyle changes break git diffs, meaning history becomes unusable (http://git.io/C5-dSQ)
- Codestyle should be consistent across the entire project for easy reading.

I can always just copy and paste your completed code and run the reformat over it once you are done though.  Your git history will already need squashed, to get this back under 3 commits.

For the record, my personal coding style has brackets on the same line.  I was originally a PHP developer.

On the functionality front, one thing that still needs done is the reimplementing of the messages, since I stripped them, while I was working on the code logic.

FarisR99 committed:
Sent messages to player.

Allows messages to be sent to the player.

FarisR99 committed:
Update messages.properties

FarisR99 committed:
Update messages_cs.properties

FarisR99 committed:
Update messages_da.properties

FarisR99 committed:
Update messages_de.properties

FarisR99 committed:
Update messages_en.properties

FarisR99 committed:
Update messages_es.properties

FarisR99 committed:
Update messages_et.properties

FarisR99 committed:
Update messages_fi.properties

FarisR99 committed:
Update messages_fr.properties

FarisR99 committed:
Update messages_hu.properties

FarisR99 committed:
Update messages_ko.properties

FarisR99 committed:
Update messages_lt.properties

FarisR99 committed:
Update messages_nl.properties

FarisR99 committed:
Update messages_pl.properties

FarisR99 committed:
Update messages_pt.properties

FarisR99 committed:
Update messages_ro.properties

FarisR99 committed:
Update messages_ru.properties

FarisR99 committed:
Update messages_sv.properties

FarisR99 committed:
Update messages_tr.properties

FarisR99 committed:
Update messages_zh.properties

FarisR99 committed:
Update messages_zh_HK.properties

FarisR99 committed:
Update messages_zh_TW.properties

FarisR99 committed:
Changed /block to /condense

Changed /block to /condense in the plugin.yml and changed the description to a more suited one. Also changed the usage.

FarisR99 committed:
Allowed NoChargeException to take in parameters.

FarisR99 committed:
Update Commandcondense.java

FarisR99 commented:
Okay, so I'm a noob at Git. I don't know how to merge all commits, etc. To edit files, I've been using the website editor. Only my own projects do I use the GitHub client. I don't know how to use it for this.

Edit: Yes please, may you format it.

khobbits commented:
https://github.com/essentials/Essentials/compare/latenight

FarisR99 commented:
I don't understand?

Edit: It seems like you've already merged them?

khobbits commented:
'latenight' is my testing branch.
I put the code I threw together last night in there, merged your changes, ran the codestyle cleanup, and rebased your changes to a single commit.

That link is a comparison of devb vs latenight branch.

If there is nothing else you want to change, I can throw this out to devb for beta testing.

FarisR99 commented:
There's nothing else I would like to change really. Thanks!
